crates/sail-python-udf/src/utils/spark.py

```diff
@@ -0,0 +1,435 @@
+from __future__ import annotations
+
+import itertools
+from typing import Any, Callable, Iterator, Union
+
+import pandas as pd
+import pyarrow as pa
+from pyspark.sql.pandas.types import _create_converter_from_pandas, _create_converter_to_pandas, from_arrow_type
+
+try:
+    _PYARROW_LIST_TYPES = (pa.ListType, pa.LargeListType, pa.FixedSizeListType, pa.ListViewType, pa.LargeListViewType)
+    _PyArrowListType = Union[pa.ListType, pa.LargeListType, pa.FixedSizeListType, pa.ListViewType, pa.LargeListViewType]
+except AttributeError:
+    _PYARROW_LIST_TYPES = (pa.ListType, pa.LargeListType, pa.FixedSizeListType)
+    _PyArrowListType = Union[pa.ListType, pa.LargeListType, pa.FixedSizeListType]
+
+try:
+    _PYARROW_LIST_ARRAY_TYPES = (
+        pa.ListArray,
+        pa.LargeListArray,
+        pa.FixedSizeListArray,
+        pa.ListViewArray,
+        pa.LargeListViewArray,
+    )
+except AttributeError:
+    _PYARROW_LIST_ARRAY_TYPES = (pa.ListArray, pa.LargeListArray, pa.FixedSizeListArray)
+
+try:
+    _PYARROW_STRING_TYPE_INSTANCES = (pa.string(), pa.large_string(), pa.string_view())
+except AttributeError:
+    _PYARROW_STRING_TYPE_INSTANCES = (pa.string(), pa.large_string())
+
+try:
+    _PYARROW_BINARY_TYPE_INSTANCES = (pa.binary(), pa.large_binary(), pa.binary_view())
+except AttributeError:
+    _PYARROW_BINARY_TYPE_INSTANCES = (pa.binary(), pa.large_binary())
+
+
+class Converter:
```

Comment:
We now have non-trivial data conversion logic that is easier to write in Python than using PyO3. This Python module is loaded once and compiled to Python byte code, so the performance is similar to calling the Python API via PyO3 in Rust.

If we ever need to further improve the performance, we could rewrite part of the code using PyO3. For now the focus is to ensure the correctness for PySpark UDF, especially in edge cases.